DOC-1753: Document the new Spelling Autocorrect plugin (#2512)
tinymce/tinymce-docs: modules/ROOT/pages/autocorrect.adoc

:navtitle: Spelling Autocorrect
:description: Autocorrect common spelling errors.
:keywords: autocorrect, capitalization, misprint, spell, spelling, spellcheck, typo, typographical error
:pluginname: Spelling Autocorrect
:plugincode: autocorrect
:pluginminimumplan: tierthree

include::partial$misc/admon-premium-plugin.adoc[]

include::partial$misc/admon-requires-6.2v.adoc[]

The {pluginname} plugin allows {productname} to autocorrect a specified set of spelling errors, common typos, and capitalization errors, as they are being typed.

== Interactive Example

liveDemo::autocorrect[]

include::partial$misc/purchase-premium-plugins.adoc[]

== Basic setup

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  plugins: 'autocorrect'
});
----

[[usage-notes]]
== Usage notes

The {pluginname} plugin currently has the following limitations:

. English is the only supported language.

. Capitalization and autocorrect can both be turned off, but specific autocorrections cannot.
+
{pluginname} can be entirely turned off, but there is, currently, no way of instructing the plugin to ignore a particular string that it considers a mis-spelling.
+

> For example, if *cna* (which the {pluginname} plugin treats as a mis-typing of *can*) is a desired string in a particular use-case, turning {pluginname} off is the currently recommended setting.

Reviewer: (discussion): use backtick or something else to make it clear that this is intentional
```suggestion
For example, if `+cna+` (which the {pluginname} plugin treats as a mis-typing of *can*) is a desired string in a particular use-case, turning {pluginname} off is the currently recommended setting.
```

> . End-users cannot, currently, turn Autocorrect on or off on hand-held devices.

Reviewer: ```suggestion
. End-users cannot, by default, turn autocorrect and capitalization on or off on hand-held devices.
```

= Spelling Autocorrect plugin
:navtitle: Spelling Autocorrect
:description: Autocorrect common spelling errors.
:keywords: autocorrect, capitalization, misprint, spell, spelling, spellcheck, typo, typographical error
:pluginname: Spelling Autocorrect
:plugincode: autocorrect
:pluginminimumplan: tierthree

include::partial$misc/admon-premium-plugin.adoc[]

> include::partial$misc/admon-requires-6.2v.adoc[]

Reviewer 2: need to add modules/ROOT/partials/misc/admon-requires-6.2v.adoc

Reviewer 3: The current path is correct.

Reviewer 3: @vpyshnenko, can you also please mark this request as addressed?

+
For example, if *cna* (which the {pluginname} plugin treats as a mis-typing of *can*) is a desired string in a particular use-case, turning {pluginname} off is the currently recommended setting.

. End-users cannot, currently, turn Autocorrect on or off on hand-held devices.

. Spelling Autocorrect is case-sensitive.
+
As a consequence, mis-spelled words are not autocorrected if they do not exactly match the plugin’s built-in correction set.
+
For example, {pluginname} autocorrects *teh* to *the*, but does not, currently, autocorrect *Teh* to *The*.

[[known-issues]]
== Known issues

The {pluginname} plugin has the following known issues:

. Autocorrect over-rides auto-capitalization.
+
As a consequence of {pluginname} being case-sensitive, if
+
  * both `autocorrect` and `capitalization` are active; and
  * a word from the specified spelling error set is mis-typed at the beginning of a sentence;
+
the spelling is corrected but the initial letter is not capitalized.

. Auto-capitalization does not work on text immediately following a soft return in non-empty {productname} documents.
+

> If
> +
>   * a {productname} document contains any data other than a soft return character (Unicode U+2028, commonly entered by typing *Shift+Return*); and
>   * the character immediately preceding an entered string is a soft return character
> +
> Auto-capitalization will not operate on the entered string.

Reviewer 4: ```suggestion
Auto-capitalization will not operate on the entered string if
+
  * a {productname} document contains any data other than a soft return character (Unicode U+2028, commonly entered by typing *Shift+Return*); and
  * the character immediately preceding an entered string is a soft return character.
+
```
I'm not sure what your preference is, just putting `If` on a single line followed by a list and then a final line looks funny to me.

. Spelling Autocorrect is case-sensitive.
+
As a consequence, mis-spelled words are not auto-corrected if they do not exactly match the plugin’s built-in correction set.
+
For example, Spelling Autocorrect auto-corrects *teh* to *the*, but does not auto-correct *Teh* to *The*.

. Spelling Autocorrect over-rides auto-capitalization.
+
As a consequence of both this and Spelling Autocorrect being case-sensitive, if
+
  * `autocorrect` and `capitalization` are both active; and
  * a word from the specified spelling error set is mis-typed at the beginning of a sentence;
+
the spelling is corrected but the initial letter is not capitalized.

. Auto-capitalization does not work on text immediately following a soft return in non-empty {productname} documents.
+
If
+
  * a {productname} document contains any data other than a soft return character (Unicode U+2028, commonly entered by typing *Shift+Return*); and
  * the character immediately preceding an entered string is a soft return character
+
Auto-capitalization will not operate on the entered string.

. Auto-capitalization interferes with media embeds.

> . Auto-capitalization interferes with `linkchecker`.

Reviewer 4: Should these be moved to the bottom of the page? I don't think the first thing people want to see is page of issues with our new plugin. Also, sounds like case-sensitivity should be a feature rather than a known issue as I doubt it needs to be "fixed". The conflict is still an issue though.

Author: > Should these be moved to the bottom of the page?

Yes.

> Also, sounds like case-sensitivity should be a feature rather than a known issue as I doubt it needs to be "fixed". The conflict is still an issue though.

To paraphrase my commit message, a bit of thinking about this lead me to *re-structure the whole document and make multiple edits and corrections*.

Reviewer 5: @brian-forte I assume the move was in relation this, but this is not how things normally are done so doing this would be inconsistent

Reviewer 5: At the very least the common/basic usage information should be at the top. As for limitations, as I said it'd be inconsistent so if you're moving one you really should be moving all. For example https://www.tiny.cloud/docs/tinymce/6/export/#client-side-pdf has the basic usage and limitations together at the top just below the example and before the options.

Author: > I assume the move was in relation this, but this is not how things normally are done so doing this would be inconsistent

Yep. My bad. Was too quick on the cut-and-paste-and-get-re-writing front. The **Options** and **Events** sections now round out the page.

Reviewer 5: That looks better thanks 👍

= Spelling Autocorrect plugin
:navtitle: Spelling Autocorrect
:description: Autocorrect common spelling mistakes.

> :keywords: autocorrect, capitalization, spell, spelling, spellcheck

Author: I’ve been thinking along the same lines. I’ve added some keywords and edited the introductory sentence accordingly.

> :keywords: autocorrect, capitalization, spell, spelling, spellcheck

Reviewer 4: Good idea, I just added a bunch of words I felt were relevant when I made this branch.

include::partial$misc/admon-requires-6.2v.adoc[]

The {pluginname} plugin allows {productname} to autocorrect a specified set of spelling errors or capitalization errors, as they are being typed.

== Interactive Example

liveDemo::autocorrect[]

include::partial$misc/purchase-premium-plugins.adoc[]

== Basic setup

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  plugins: 'autocorrect'
});
----

[[usage-notes]]
== Usage notes

The {pluginname} plugin currently has the following limitations:

. English is the only supported language.

. Capitalization and autocorrect can both be turned off, but specific autocorrections cannot.
+
{pluginname} can be entirely turned off, but there is, currently, no way of instructing the plugin to ignore a particular string that it considers a mis-spelling.
+
For example, if *cna* (which the {pluginname} plugin treats as a mis-typing of *can*) is a desired string in a particular use-case, turning {pluginname} off is the currently recommended setting.

. End-users cannot, currently, turn Autocorrect on or off on hand-held devices.

. Spelling Autocorrect is case-sensitive.
+

> As a consequence, mis-spelled words are not auto-corrected if they do not exactly match the plugin’s built-in correction set.

Author: > The hyphenation here is a llttle bit inconsistent, but I'm not sure if it matters. We have:
> 
> autocorrect, auto-corrected, auto-correct, auto-corrects

Good catch. Doing some quick digging, _autocorrect_ emerged entirely from technical en-US. Specifically from folks producing software akin to this plugin.

And, as is somewhat common when en-US speakers coin new words by compounding, it never took a hyphen. (en-US is notorious for getting rid of hyphens quickly, especially compared to other Englishes, especially en-IN.)

FWIW, the sort-of-jokey-but-perhaps-not-entirely-without-merit hypothesis among linguists is that German speakers had a bigger influence on normative en-US spelling norms than is commonly realised. Compounding in German virtually never uses the hyphen glyph.

> As a consequence, mis-spelled words are not auto-corrected if they do not exactly match the plugin’s built-in correction set.

Author: > We also use spelling mistakes in one place, and spelling errors in another. Really minor, I'm just wondering if it's intentional.

Another good catch.

I suspect the proximate source of the inconsistency is the document being put together by _[Dyvers Hands](https://en.wikipedia.org/wiki/Divers_hands)_.

But, on reflection, I think _error_ is the better term here. In common use, _mistake_ has connotations of intent and blame. Someone makes a mistake and things being wrong are, therefore, their fault.

By contrast, _error_ is normally used and read without such connotations. The thing is still in the wrong state, and it’s entirely likely the wrong state is consequent to someone’s action. But you aren’t going to get blamed for an _error_, even if you might also be responsible for correcting it.

Typos, in particular, feel like errors, not mistakes, to me. (And, yes, adding the noun phrase `typographical errors` to the keywords runs counter to this. But `typographical error` is the fixed phrase, although it’s now almost completely replaced by `typo`, even in the most formal of editorial environments.)

+
For example, {pluginname} auto-corrects *teh* to *the*, but does not, currently, auto-correct *Teh* to *The*.

[[known-issues]]
== Known issues

The {pluginname} plugin has the following known issues:

> . {pluginname} over-rides auto-capitalization.

Reviewer 4: I agree this should simply be `autocorrect`

+
As a consequence of {pluginname} being case-sensitive, if
+
  * both `autocorrect` and `capitalization` are active; and
  * a word from the specified spelling error set is mis-typed at the beginning of a sentence;
+
the spelling is corrected but the initial letter is not capitalized.

. Switching to a language other than English disables the Autocorrect and Capitalization menus
+
When the `autocorrect` plugin loads, *Autocorrect* and *Capitalization* menu items are added to the *Tools* menu.
+

> If {productname} is then switched to a language other than English, both these menu items are disabled. And they are not automatically re-enabled if a user switches back to English.

Author: The far more accurate NOTE admonition re interaction between Spell Checker Pro and Autocorrect is now a partial in its own right.

And `include` statements grabbing said partial are now in `autocorrect_autocorrect.adoc`, `autocorrect_capitalization.adoc`, *and* `autocorrect.adoc`.

> . Auto-capitalization does not work on text immediately following a soft return in non-empty {productname} documents.

Author: Call this a combination of laziness and _I was a SysAdmin for several years_ twitchiness.

1. The laziness
what’s in the docs is a very lightly edited version of what I put down as I was, quite literally, discovering the behaviour while I was going through all the queries directed at me in @ShiridiGandham’s [QA spreadsheet](https://docs.google.com/spreadsheets/d/1YKuFZ-xvUDmFBq4kh96ZiRVy5La2NajVvsbmIuSskNk/edit?usp=sharing). It was quicker to lightly-edit than re-write.
2. The twitchiness
When bugs or known issues are not explicitly and fully documented, I’m immediately brought to mind of way too many times I was at a console at 03:00 trying to work out why my [application others are relying on here] wasn’t working and the notes accompanying the [will this patch make everything better?] being maddeningly vague.

+
If
+
  * a {productname} document contains any data other than a soft return character (Unicode U+2028, commonly entered by typing *Shift+Return*); and
  * the character immediately preceding an entered string is a soft return character
+
Auto-capitalization will not operate on the entered string.

> . Auto-capitalization interferes with media embeds.

Reviewer 6: How about - "Auto-capitalization interferes with Enhanced Media Embed"

> . Auto-capitalization interferes with media embeds.

Reviewer 5: Yeah using `media embeds` here would be confusing, as that will also mean media inserted via the `media` plugin and this only interferes with Enhanced Media as-you-type functionality IIUC.

> . Auto-capitalization interferes with media embeds.

Author: Both one-liners now edited.

FYI: _interferes with_ is maddeningly vague to me. But I’ve no info as to the nature of the interference.
